### ISSUE TITLE
[Bugs?] PFCP Session Establishment failed with free5GC v3.3.0 and oai-cn5g-upf-vpp v1.5.1

### ISSUE THREAD

**Issue by s5uishida** [free5gc/free5gc#457]
## Describe the bug
`PFCP Session Establishment` failed with free5GC v3.3.0 and [oai-cn5g-upf-vpp](https://gitlab.eurecom.fr/oai/cn5g/oai-cn5g-upf-vpp) v1.5.1.
I don't know if this is a bug or not.

## To Reproduce
`PFCP Session Establishment` was successful with free5GC v3.2.1 (not v3.3.0) and oai-cn5g-upf-vpp v1.5.1.
I wrote this successful configuration as folows.

https://github.com/s5uishida/free5gc_ueransim_vpp_upf_dpdk_sample_config

However, when I used this configuration as is, and changed free5GC from v3.2.1 to v3.3.0 and tried to connect the UE, I received the following error message (pcap) in the `PFCP Session Establishment Response`.

```
    Error Report : [Grouped IE]
        Enterprise specific IE Type: 32774
        IE Length: 42
        Enterprise ID: Travelping GmbH (18681)
-->     Error Message : PDR ID 1, unknown Network Instance
            Enterprise specific IE Type: 32775
            IE Length: 36
            Enterprise ID: Travelping GmbH (18681)
            Error Message: PDR ID 1, unknown Network Instance
```
For DNN and Network Instance, I am using `internet`.
IMSI is `001010000000000`.

## Expected behavior
`PFCP Session Establishment` succeeds with free5GC v3.3.0 and oai-cn5g-upf-vpp v1.5.1.

## Environment (please complete the following information):
 - free5GC Version: [v3.3.0]
 - OS: [Ubuntu 22.04 Server]
 - Kernel version: [5.15.0-67-lowlatency]
 - go version: [go1.17.8 linux/amd64]
 - c compiler version (Option): [gcc version 11.3.0]

### Configuration File
The changed configuration files are as follows.
- amfcfg.yaml
```
--- amfcfg.yaml.orig    2023-06-17 13:32:30.936184528 +0900
+++ amfcfg.yaml 2023-06-17 13:43:54.901214657 +0900
@@ -5,7 +5,7 @@
 configuration:
   amfName: AMF # the name of this AMF
   ngapIpList:  # the IP list of N2 interfaces on this AMF
-    - 127.0.0.18
+    - 192.168.0.141
   ngapPort: 38412 # the SCTP port listened by NGAP
   sbi: # Service-based interface information
     scheme: http # the protocol for sbi (http or https)
@@ -24,18 +24,18 @@
   servedGuamiList: # Guami (Globally Unique AMF ID) list supported by this AMF
     # <GUAMI> = <MCC><MNC><AMF ID>
     - plmnId: # Public Land Mobile Network ID, <PLMN ID> = <MCC><MNC>
-        mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-        mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+        mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+        mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
       amfId: cafe00 # AMF identifier (3 bytes hex string, range: 000000~FFFFFF)
   supportTaiList:  # the TAI (Tracking Area Identifier) list supported by this AMF
     - plmnId: # Public Land Mobile Network ID, <PLMN ID> = <MCC><MNC>
-        mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-        mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+        mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+        mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
       tac: 000001 # Tracking Area Code (3 bytes hex string, range: 000000~FFFFFF)
   plmnSupportList: # the PLMNs (Public land mobile network) list supported by this AMF
     - plmnId: # Public Land Mobile Network ID, <PLMN ID> = <MCC><MNC>
-        mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-        mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+        mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+        mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
       snssaiList: # the S-NSSAI (Single Network Slice Selection Assistance Information) list supported by this AMF
         - sst: 1 # Slice/Service Type (uinteger, range: 0~255)
           sd: 010203 # Slice Differentiator (3 bytes hex string, range: 000000~FFFFFF)
```
- ausfcfg.yaml
```
--- ausfcfg.yaml.orig   2023-06-17 13:32:30.937184630 +0900
+++ ausfcfg.yaml        2023-06-17 13:44:45.741848358 +0900
@@ -15,8 +15,8 @@
     - nausf-auth # Nausf_UEAuthentication service
   nrfUri: http://127.0.0.10:8000 # a valid URI of NRF
   plmnSupportList: # the PLMNs (Public Land Mobile Network) list supported by this AUSF
-    - mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-      mnc: 93  # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+    - mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+      mnc: 01  # Mobile Network Code (2 or 3 digits string, digit: 0~9)
     - mcc: 123 # Mobile Country Code (3 digits string, digit: 0~9)
       mnc: 45  # Mobile Network Code (2 or 3 digits string, digit: 0~9)
   groupId: ausfGroup001 # ID for the group of the AUSF
```
- nrfcfg.yaml
```
--- nrfcfg.yaml.orig    2023-06-17 13:32:30.937184630 +0900
+++ nrfcfg.yaml 2023-06-17 13:45:55.575758415 +0900
@@ -14,8 +14,8 @@
       pem: cert/nrf.pem # NRF TLS Certificate
       key: cert/nrf.key # NRF TLS Private key
   DefaultPlmnId:
-    mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-    mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+    mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+    mnc: 001 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
   serviceNameList: # the SBI services provided by this NRF, refer to TS 29.510
     - nnrf-nfm # Nnrf_NFManagement service
     - nnrf-disc # Nnrf_NFDiscovery service
```
- nssfcfg.yaml
```
--- nssfcfg.yaml.orig   2023-06-17 13:32:30.937184630 +0900
+++ nssfcfg.yaml        2023-06-17 13:46:37.986325676 +0900
@@ -17,12 +17,12 @@
     - nnssf-nssaiavailability # Nnssf_NSSAIAvailability service
   nrfUri: http://127.0.0.10:8000 # a valid URI of NRF
   supportedPlmnList: # the PLMNs (Public land mobile network) list supported by this NSSF
-    - mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-      mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+    - mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+      mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
   supportedNssaiInPlmnList: # Supported S-NSSAI List for each PLMN
     - plmnId: # Public Land Mobile Network ID, <PLMN ID> = <MCC><MNC>
-        mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-        mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+        mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+        mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
       supportedSnssaiList: # Supported S-NSSAIs of the PLMN
         - sst: 1 # Slice/Service Type (uinteger, range: 0~255)
           sd: 010203 # Slice Differentiator (3 bytes hex string, range: 000000~FFFFFF)
```
- smfcfg.yaml
```
--- smfcfg.yaml.orig    2023-06-17 13:32:30.937184630 +0900
+++ smfcfg.yaml 2023-06-17 13:53:16.643854000 +0900
@@ -34,22 +34,22 @@
             ipv4: 8.8.8.8
             ipv6: 2001:4860:4860::8888
   plmnList: # the list of PLMN IDs that this SMF belongs to (optional, remove this key when unnecessary)
-    - mcc: 208 # Mobile Country Code (3 digits string, digit: 0~9)
-      mnc: 93 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
+    - mcc: 001 # Mobile Country Code (3 digits string, digit: 0~9)
+      mnc: 01 # Mobile Network Code (2 or 3 digits string, digit: 0~9)
   locality: area1 # Name of the location where a set of AMF, SMF, PCF and UPFs are located
   pfcp: # the IP address of N4 interface on this SMF (PFCP)
     # addr config is deprecated in smf config v1.0.3, please use the following config
-    nodeID: 127.0.0.1 # the Node ID of this SMF
-    listenAddr: 127.0.0.1 # the IP/FQDN of N4 interface on this SMF (PFCP)
-    externalAddr: 127.0.0.1 # the IP/FQDN of N4 interface on this SMF (PFCP)
+    nodeID: 192.168.14.141 # the Node ID of this SMF
+    listenAddr: 192.168.14.141 # the IP/FQDN of N4 interface on this SMF (PFCP)
+    externalAddr: 192.168.14.141 # the IP/FQDN of N4 interface on this SMF (PFCP)
   userplaneInformation: # list of userplane information
     upNodes: # information of userplane node (AN or UPF)
       gNB1: # the name of the node
         type: AN # the type of the node (AN or UPF)
       UPF: # the name of the node
         type: UPF # the type of the node (AN or UPF)
-        nodeID: 127.0.0.8 # the Node ID of this UPF
-        addr: 127.0.0.8 # the IP/FQDN of N4 interface on this UPF (PFCP)
+        nodeID: 192.168.14.151 # the Node ID of this UPF
+        addr: 192.168.14.151 # the IP/FQDN of N4 interface on this UPF (PFCP)
         sNssaiUpfInfos: # S-NSSAI information list for this UPF
           - sNssai: # S-NSSAI (Single Network Slice Selection Assistance Information)
               sst: 1 # Slice/Service Type (uinteger, range: 0~255)
@@ -72,7 +72,7 @@
         interfaces: # Interface list for this UPF
           - interfaceType: N3 # the type of the interface (N3 or N9)
             endpoints: # the IP address of this N3/N9 interface on this UPF
-              - 127.0.0.8
+              - 192.168.13.151
             networkInstances:  # Data Network Name (DNN)
               - internet
     links: # the topology graph of userplane, A and B represent the two nodes of each link
@@ -91,6 +91,7 @@
   nrfUri: http://127.0.0.10:8000 # a valid URI of NRF
   #urrPeriod: 10 # default usage report period in seconds
   #urrThreshold: 1000 # default usage report threshold in bytes
+  ulcl: false
 
 logger: # log output setting
   enable: true # true or false
```
I attached the entire config files just in case.
[config.zip](https://github.com/free5gc/free5gc/files/11799723/config.zip)

### PCAP File
[0620_02.zip](https://github.com/free5gc/free5gc/files/11799854/0620_02.zip)

### Log File
```
2023-06-20T19:26:52.848850008+09:00 [INFO][SMF][PduSess] Sending PFCP Session Establishment Request
2023-06-20T19:26:52.849866879+09:00 [INFO][AMF][Gmm][amf_ue_ngap_id:RU:1,AU:1(3GPP)][supi:SUPI:imsi-001010000000000] create smContext[pduSessionID: 1] Success
2023-06-20T19:26:52.852344722+09:00 [INFO][SMF][PduSess] Received PFCP Session Establishment Not Accepted Response
```
- **PFCP Session Establishment Request (pcap)**
```
Frame 1397: 379 bytes on wire (3032 bits), 379 bytes captured (3032 bits)
Linux cooked capture v2
Internet Protocol Version 4, Src: 192.168.14.141, Dst: 192.168.14.151
User Datagram Protocol, Src Port: 8805, Dst Port: 8805
Packet Forwarding Control Protocol
    Flags: 0x23, Message Priority (MP), SEID (S)
    Message Type: PFCP Session Establishment Request (50)
    Length: 327
    SEID: 0x0000000000000000
    Sequence Number: 2
    .... .... .... .... 0000 .... = Message Priority: 0
    .... 0000 = Spare: 0
    Node ID : IPv4 address: 192.168.14.141
    F-SEID : SEID: 0x0000000000000001, IPv4 192.168.14.141
    Create PDR : [Grouped IE]: PDR ID: 1
        IE Type: Create PDR (1)
        IE Length: 86
        PDR ID : 1
        Precedence : 255
        PDI : [Grouped IE]
            IE Type: PDI (2)
            IE Length: 39
            Source Interface : Access
            F-TEID : TEID: 0x00000001, IPv4 192.168.13.151
-->         Network Instance : internet
                IE Type: Network Instance (22)
                IE Length: 8
                Network Instance: internet
            UE IP Address : 
        Outer Header Removal : GTP-U/UDP/IPv4
        FAR ID : Dynamic by CP 1
        QER ID : Dynamic by CP 2
        QER ID : Dynamic by CP 1
    Create PDR : [Grouped IE]: PDR ID: 2
    Create FAR : [Grouped IE]: FAR ID: Dynamic by CP 1
    Create FAR : [Grouped IE]: FAR ID: Dynamic by CP 2
    Create QER : [Grouped IE]: QER ID: Dynamic by CP 2
    Create QER : [Grouped IE]: QER ID: Dynamic by CP 1
    PDN Type : IPv4
    [Response In: 1404]
```
- **PFCP Session Establishment Response (pcap)**
```
Frame 1404: 159 bytes on wire (1272 bits), 159 bytes captured (1272 bits)
Linux cooked capture v2
Internet Protocol Version 4, Src: 192.168.14.151, Dst: 192.168.14.141
User Datagram Protocol, Src Port: 8805, Dst Port: 8805
Packet Forwarding Control Protocol
    Flags: 0x21, SEID (S)
    Message Type: PFCP Session Establishment Response (51)
    Length: 107
    SEID: 0x0000000000000001
    Sequence Number: 2
    Spare: 0
    Node ID : FQDN: 192.168.14.151
        IE Type: Node ID (60)
        IE Length: 16
        0000 .... = Spare: 0
        .... 0010 = Address Type: FQDN (2)
        FQDN: 192.168.14.151
    Cause : Rule creation / modification Failure
        IE Type: Cause (19)
        IE Length: 1
        Cause: Rule creation / modification Failure (73)
    Error Report : [Grouped IE]
        Enterprise specific IE Type: 32774
        IE Length: 42
        Enterprise ID: Travelping GmbH (18681)
-->     Error Message : PDR ID 1, unknown Network Instance
            Enterprise specific IE Type: 32775
            IE Length: 36
            Enterprise ID: Travelping GmbH (18681)
            Error Message: PDR ID 1, unknown Network Instance
    F-SEID : SEID: 0x0000000000000001, IPv4 192.168.14.151
        IE Type: F-SEID (57)
        IE Length: 13
        0... .... = Spare: 0
        .0.. .... = Spare: 0
        ..0. .... = Spare: 0
        ...0 .... = Spare: 0
        .... 0... = Spare: 0
        .... .0.. = Spare: 0
        .... ..1. = V4 (IPv4): Present
        .... ...0 = V6 (IPv6): Not Present
        SEID: 0x0000000000000001
        IPv4 address: 192.168.14.151
    Failed Rule ID : PDR: 1
        IE Type: Failed Rule ID (114)
        IE Length: 3
        .... .000 = Failed Rule ID Type: PDR (0)
        Rule ID: 1
    [Response To: 1397]
    [Response Time: 0.003092000 seconds]
```
Any idea why `PFCP Session Establishment` is successful in free5GC v3.2.1 but fails in v3.3.0?

Best regards,

--Shigeru

## Additional context

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Would you please help to provide the log of SMF?

Thank you.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Thank you for your reply.

I can't work on this right now, so I'll deal with it later.

BTW, could you please let me know a little?

Is there a way to create a log file for each NF?
I will get a trace level log for SMF, but I couldn't find a place to specify the output destination of the log file dedicated to SMF.
If it is not possible to create a log file for each NF, the stdout of all C-Plane NFs is captured as a single log file.

Regards,

--Shigeru

**Comment by ianchen0119** [comment]
Hi @s5uishida 
Maybe you can use docker to launch the entire 5GC service instead of the host deployment?
In this way,
You can use the command `docker logs smf` to get the log of specific service.

Thanks.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

I got the log files for each NF and the trace log for SMF.
I attached these log files along with the pcap file below.

[20230621-1.zip](https://github.com/free5gc/free5gc/files/11817876/20230621-1.zip)

The config files are the same as these I attached earlier.

Thank you very much in advance.

Best regards,

--Shigeru

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Please add the `nwInstFqdnEncoding: true` attribute in `smfcfg.yaml`, it will look like:
```yaml=
configuration:
  # ...
  nwInstFqdnEncoding: true
```

Thanks.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Thank you very much for your reply!

By adding this setting, `Network Instance` of `PFCP Session Establishment Request` has `IE Length: 9` (not  `IE Length: 8`) as shown by the arrow below, and the successful Response was returned.

```
Create PDR : [Grouped IE]: PDR ID: 1
    IE Type: Create PDR (1)
    IE Length: 87
    PDR ID : 1
    Precedence : 255
    PDI : [Grouped IE]
        IE Type: PDI (2)
        IE Length: 40
        Source Interface : Access
        F-TEID : TEID: 0x00000001, IPv4 192.168.13.151
        Network Instance : internet
            IE Type: Network Instance (22)
-->         IE Length: 9
            Network Instance: internet
        UE IP Address : 
    Outer Header Removal : GTP-U/UDP/IPv4
    FAR ID : Dynamic by CP 1
    QER ID : Dynamic by CP 2
    QER ID : Dynamic by CP 1
```
I attach the pcap file at this time.

[nwInstFqdnEncoding_true_0624_02.zip](https://github.com/free5gc/free5gc/files/11851641/nwInstFqdnEncoding_true_0624_02.zip)

I think that the commit of `add FQDNEncoding option to switch NetworkInstance encoding` is effectively disabled.

https://github.com/free5gc/pfcp/commit/18a0d89001aee2d56513c9a3fbadae076169a29e

However, the Data Network that was created turned out to be strange. Specifically, when I ran `tcpdump` on the N6 of the Data Network Gateway VM and pinged `google.com` from the UE, the following log appeared and the UE did not receive the ping response.

On Data Network Gateway VM)
```
# tcpdump -i enp0s9 -n
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on enp0s9, link-type EN10MB (Ethernet), snapshot length 262144 bytes
03:52:46.764852 ARP, Request who-has 192.168.16.152 tell 192.168.16.151, length 46
03:52:46.764889 ARP, Reply 192.168.16.152 is-at 08:00:27:66:03:40, length 28
03:52:47.772878 IP 10.60.0.1 > 142.250.207.14: ICMP echo request, id 1, seq 2, length 64
03:52:47.790110 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790215 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790223 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790288 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790292 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790365 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790370 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790420 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790423 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790473 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790476 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
...
```
As I wrote in my configuration log, free5GC v3.2.1 (updated 2023.02.12) works fine without problems. I suspect it's caused by code committed between 2023.02.12 and v.3.3.0.

In addition, if it is UPF of v3.3.0, it will work regardless of whether `nwInstFqdnEncoding: true` is set or not, but I don't think it will work with other UPFs as it is.

Sorry if I misunderstood.

If you find out anything, it would be very glad if you could let me know.

Thank you very much in advance!

Best regards,

--Shigeru

**Comment by ianchen0119** [comment]
Hi @s5uishida 

It's happy to see that the session establishment works now!

> However, the Data Network that was created turned out to be strange. Specifically, when I ran tcpdump on the N6 of the Data Network Gateway VM and pinged google.com from the UE, the following log appeared and the UE did not receive the ping response.

According to the information you gave, 
I think the UE has been allocated an IP address (10.60.0.1) correctly by SMF, and its ICMP echo request has been replied from `google.com`.
So, I think the problem is that UPF did not forward the downlink packets to the N3 interface, Could you help to know which UPF implementation you used in this deployment?

> In addition, if it is UPF of v3.3.0, it will work regardless of whether nwInstFqdnEncoding: true is set or not, but I don't think it will work with other UPFs as it is.

I think you are right! I will update the default SMF configuration to enable the `nwInstFqdnEncoding`.



Thank you!

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Thank you very much for your reply.

> So, I think the problem is that UPF did not forward the downlink packets to the N3 interface, Could you help to know which UPF implementation you used in this deployment?

As reported here, the UPF which I'm using is [oai-cn5g-upf-vpp](https://gitlab.eurecom.fr/oai/cn5g/oai-cn5g-upf-vpp) v1.5.1.

I expect VPP-UPF with DPDK to speed up the Data-Plane. In my configuration log, I'm using a Virtualbox VM, but in reality, I'm assuming to use a server with physical NICs.

`ping` works fine with this UPF and free5GC v3.2.1 (updated 2023.02.12). On the other hand, `ping` does not work with the combination of this UPF and free5GC v3.3.0.

> I think you are right! I will update the default SMF configuration to enable the `nwInstFqdnEncoding`.

Is the committed code for enable or disable `nwInstFqdnEncoding` the correct implementation?

It says to enable the default `nwInstFqdnEncoding`, but when should `nwInstFqdnEncoding` be disable?

Thank you very much in advance!

**Comment by s5uishida** [comment]
Hi @infinitydon

Sorry for tagging you directly.

Please let me know if you know.

Does Data-Plane work fine with free5GC v3.3.0 and upg-vpp?

Thank you very much in advance.

**Comment by infinitydon** [comment]
@s5uishida I haven't tested it yet, the last time I tried I think there's need to modify free5gc SMF code before it can work but this is more 1 year ago that I tried it.

**Comment by s5uishida** [comment]
@infinitydon 

OK, thanks.

When try it with free5GC v3.3.0, I would be very glad if you could share the result.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Regarding `nwInstFqdnEncoding`, I confirmed the meaning of the function from the following specifications.

Thank you very much!
```
3GPP TS 29.244 LTE;5G;Interface between the Control Plane and the User Plane nodes
- 8 Information Elements
  - 8.2 Information Elements
    - 8.2.4 Network Instance

3GPP TS 23.003 LTE;5G;Numbering, addressing and identification
- 9A Definition of Data Network Name
  - 9.1 Structure of APN

The encoding of the APN shall follow the Name Syntax defined in RFC 2181 [18], RFC 1035 [19] and RFC 1123 [20].
The APN consists of one or more labels. Each label is coded as a one octet length field followed by that number of
octets coded as 8 bit ASCII characters. Following RFC 1035 [19] the labels shall consist only of the alphabetic
characters (A-Z and a-z), digits (0-9) and the hyphen (-). Following RFC 1123 [20], the label shall begin and end with
either an alphabetic character or a digit. The case of alphabetic characters is not significant. The APN is not terminated
by a length byte of zero.
```
I also think that `nwInstFqdnEncoding: true` is good by default in `smfcfg.yaml`.

However, the issue of not being able to receive a ping response on the UE has not been resolved yet.

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Thanks for your information, I would have a discussion with my teammate about changing the default configuration.

Let's go back to the ping issue,
I saw the network traffic on your data network gateway:
```
# tcpdump -i enp0s9 -n
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on enp0s9, link-type EN10MB (Ethernet), snapshot length 262144 bytes
03:52:46.764852 ARP, Request who-has 192.168.16.152 tell 192.168.16.151, length 46
03:52:46.764889 ARP, Reply 192.168.16.152 is-at 08:00:27:66:03:40, length 28
03:52:47.772878 IP 10.60.0.1 > 142.250.207.14: ICMP echo request, id 1, seq 2, length 64
03:52:47.790110 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790215 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790223 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790288 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790292 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790365 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790370 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790420 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790423 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790473 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
03:52:47.790476 IP 142.250.207.14 > 10.60.0.1: ICMP echo reply, id 1, seq 2, length 64
...
```
It seems like there is no NAT for the UE's IP, I think the data network gateway hasn't a correct routing rule to forward the user plane data.


Thanks.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Data Network Gateway settings are as described [here](https://github.com/s5uishida/install_vpp_upf_dpdk#setup-data-network-gateway-on-vm-dn).

On this DN/GW, NAPT and routing to the UE IP address are set as follows.
```
# iptables -t nat -A POSTROUTING -s 10.60.0.0/16 -j MASQUERADE
# ip route add 10.60.0.0/16 via 192.168.16.151 dev enp0s9
```
`192.168.16.151` is the N6 IP address of VPP-UPF (oai-cn5g-upf-vpp).

Are there any other settings required?

For free5GC v3.2.1, with this setting the UE was able to receive ping responses.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

As you can see from the URL of my configuration, I think I have included all the information necessary to reproduce my environment.

So if you don't mind, I think you could build the similar environment. In addition, I installed Virtualbox in one PC and builded all environments with this one.

Even though I have the necessary network settings for DN/GW, it's very strange that the ping is successful with free5GC v3.2.1 and this failure behavior with v3.3.0.

FYI.

Best regards,

--Shigeru

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Please checkout the branch of SMF to `bugfix/gh-issue-457` and retry again:
```
cd smf
git fetch origin 
git checkout bugfix/gh-issue-457
```

Thank you.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

`bugfix/gh-issue-457` had the following errors. The correct diff from v3.3.0 is as follows.

1) `~/free5gc/NFs/smf/internal/sbi/pdusession/routers.go`

```
--- routers.go.orig     2023-06-23 10:58:50.286792990 +0900
+++ routers.go  2023-06-26 20:02:54.353171027 +0900
@@ -119,6 +119,6 @@
                "PostSmContexts",
                strings.ToUpper("Post"),
                "/sm-contexts",
-               HTTPPostSmContexts,
+               SHTTPPostSmContexts,
        },
 }
```

2) `~/free5gc/NFs/smf/internal/sbi/producer/pdu_session.go`

No changes from v3.3.0 are required.

Now ping works fine. Also, I was able to download the file with the curl command via the TUNnel interface (IP address = `10.60.0.1`) generated on the UE side (UERANSIM).

Thank you very much!!

I look forward to the release of v3.3.1. When v3.3.1 is released, I'll update my configuration URL as well.

Best regards,

--Shigeru

**Comment by s5uishida** [comment]
Hi @ianchen0119 

I will correct.

In `bugfix/gh-issue-457`, only the following fixes were necessary.

`free5gc/NFs/smf/internal/context/datapath.go`
```
--- datapath.go.orig    2023-06-26 20:56:34.786119308 +0900
+++ datapath.go 2023-06-26 20:29:40.310824497 +0900
@@ -584,7 +584,6 @@
                        DLPDR.Precedence = precedence
 
                        // TODO: Should delete this after FR5GC-1029 is solved
-                       // TODO: Should delete this after FR5GC-1029 is solved
                        if curDataPathNode.IsAnchorUPF() {
                                DLPDR.PDI = PDI{
                                        SourceInterface: pfcpType.SourceInterface{
@@ -679,22 +678,6 @@
                                }
                        }
                }
-               if curDataPathNode.DownLinkTunnel != nil {
-                       if curDataPathNode.DownLinkTunnel.SrcEndPoint == nil {
-                               DNDLPDR := curDataPathNode.DownLinkTunnel.PDR
-                               DNDLPDR.PDI = PDI{
-                                       SourceInterface: pfcpType.SourceInterface{InterfaceValue: pfcpType.SourceInterfaceCore},
-                                       NetworkInstance: &pfcpType.NetworkInstance{
-                                               NetworkInstance: smContext.Dnn,
-                                               FQDNEncoding:    factory.SmfConfig.Configuration.NwInstFqdnEncoding,
-                                       },
-                                       UEIPAddress: &pfcpType.UEIPAddress{
-                                               V4:          true,
-                                               Ipv4Address: smContext.PDUAddress.To4(),
-                                       },
-                               }
-                       }
-               }
        }
 
        dataPath.Activated = true
```
No modification of the other 3 files was required.

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Yes, your change is correct.
Actually, I have only changed the correct one, any other files were changed by mistake😅

Thanks.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Please close this issue once this fix is merged into SMF.

Thank you very much for your contribution!

**Comment by ianchen0119** [comment]
Hi @s5uishida 

No problem!
BTW:
I think your sample configuration is useful to anyone who is interested in vpp-upf integration.
So I'm wondering If you agree that we embed your repository link to our [official website](https://free5gc.org/blog/).

Thanks!

**Comment by s5uishida** [comment]
Hi @ianchen0119 

Sure!

I would be very glad if you could link to my repository.

Thanks!

**Comment by ianchen0119** [comment]
Hi @s5uishida 

Your article has been added to our website, you can check out the link below:
https://free5gc.org/blog/

Thanks a lot.

**Comment by s5uishida** [comment]
Hi @ianchen0119 

I think it would be good to add the following in `smfcfg.yaml` by default.
```
  nwInstFqdnEncoding: true
```
Thanks.
